python/hyperonpy.cpp

```diff
@@ -330,6 +339,56 @@ PYBIND11_MODULE(hyperonpy, m) {
         return var_atom_list;
     }, "Returns iterator to traverse bindings");
 
+    py::class_<CBindingsSet>(m, "CBindingsSet");
+    m.def("bindings_set_empty", []() { return CBindingsSet(bindings_set_empty()); }, "New BindingsSet with no Bindings");
+    m.def("bindings_set_single", []() { return CBindingsSet(bindings_set_single()); }, "New BindingsSet with one new Bindings");
+    m.def("bindings_set_free", [](CBindingsSet set) { bindings_set_free(set.ptr); }, "Free BindingsSet");
+    m.def("bindings_set_eq", [](CBindingsSet set, CBindingsSet other) { return bindings_set_eq(set.ptr, other.ptr); }, "Free BindingsSet");
+    //TODO: Discussion topic.  I would like to replace BindingsSet::len() with BindingsSet::is_empty across the whole
+    // code base, all the way back to Rust. Then I'd like to unify the behavior between a BindingsSet with a single empty
+    // Bindings and a BindingsSet with no Bindings whatsoever.  I believe this is one step towards a future where
+    // BindingsSets can represent lazily-evaluated bindings, and ultimately infinite and explosive sets
+    m.def("bindings_set_is_empty", [](CBindingsSet set) {
+        if (bindings_set_len(set.ptr) == 0) return true;
+        bindings_set_t* test_set = bindings_set_single();
+        bool val = bindings_set_eq(set.ptr, test_set);
+        bindings_set_free(test_set);
+        return val;
+    }, "Returns true if BindingsSet contains no Bindings");
```

Comment:
1. I would wait for the use-case in code first. The current logic doesn't treat `BindingsSet` like a better version of `Bindings`. It treats it as a set of `Bindings` and always goes through this set to work with each `Bindings` instance. For instance it doesn't make sense to introduce `BindingsSet::resolve()` because all of the variables values in some `Bindings` instance form the inference context and one value is not useful out of this context. `BindingsSet::single` in matcher means that two atoms are equal, and `BindingsSet::empty` means they cannot be matched. Thus I don't see the use-case for such semantics of `_empty()` yet. And if this semantics is not used then name can be confusing because one may think it returns true iff no `Bindings` instances are inside.

2. If you are agree with first comment we have two options here: introduce `_len` function in `hyperonpy.cpp` and map it to the C API function, or leave `_empty` in `hyperonpy.cpp` then probably it makes sense to replace `_len` by `_empty` in C API.

Reply:
There is a conceptual subtlety here that I don't fully appreciate.  My preference is to eliminate bindings_set_len() from the C API (and also from Rust), but you raise an interesting question about how to express (in code and also in documentation) the difference between a BindingsSet with one empty Bindings, and a BindingsSet with no Bindings.

As you say, one is the absence of constraints, and the other is fully constrained.  I think another word is needed, as opposed to empty.

Reply:
I filled this issue to keep the discussion in one place:  https://github.com/trueagi-io/hyperon-experimental/issues/281

Reply:
In my opinion this PR can be merged - even without finalized names for empty & single.  It will be a simple matter to rename them.  And I would hate to hold back other work on a conceptual discussion that might take longer to come to a conclusion.